sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,633 @@
+		ret = -ENOMEM;
+		goto err_in;
+	}
+
+	ret = sof_update_ipc_object(scomp, out_format,
+				    SOF_IPC4_OUT_AUDIO_FORMAT_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(*out_format),
+				    available_fmt->audio_fmt_num);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse output audio_fmt tokens failed\n");
+		goto err_out;
+	}
+
+	available_fmt->out_audio_fmt = out_format;
+	sof_ipc4_dbg_audio_format(scomp->dev, out_format, sizeof(*out_format),
+				  available_fmt->audio_fmt_num,
+				  swidget->widget->name, __func__);
+
+finish:
+	return 0;
+
+err_out:
+	kfree(out_format);
+err_in:
+	kfree(base_config);
+
+	return ret;
+}
+
+
+static void sof_ipc4_widget_free_comp(struct snd_sof_widget *swidget)
+{
+	kfree(swidget->private);
+}
+
+static int sof_ipc4_widget_setup_pcm(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_host *host;
+	int node_type = 0;
+	int ret;
+
+	host = kzalloc(sizeof(*host), GFP_KERNEL);
+	if (!host)
+		return -ENOMEM;
+
+	swidget->private = host;
+
+	ret = sof_ipc4_get_audio_fmt(scomp, swidget, &host->available_fmt, true);
+	if (ret != 0)
+		goto err;
+
+	ret = sof_update_ipc_object(scomp, &host->copier.gtw_cfg,
+				    SOF_IPC4_COPIER_GATEWAY_CFG_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(host->copier.gtw_cfg), 1);
+
+	if (ret != 0 || !host->copier.gtw_cfg.dma_buffer_size) {
+		dev_err(scomp->dev, "Invalid host dma_buffer_size %d\n",
+			host->copier.gtw_cfg.dma_buffer_size);
+		goto err;
+	}
+	dev_dbg(scomp->dev, "host %s gtw_cfg dma_buffer_size %u\n",
+		swidget->widget->name, host->copier.gtw_cfg.dma_buffer_size);
+
+	ret = sof_update_ipc_object(scomp, &node_type,
+				    SOF_IPC4_COPIER_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(node_type), 1);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse host copier_node tokens failed %d\n",
+			ret);
+		goto err;
+	}
+	dev_dbg(scomp->dev, "host %s node_type %u\n",
+		swidget->widget->name, node_type);
+
+	host->copier.gtw_cfg.node_id = SOF_IPC4_NODE_TYPE(node_type);
+
+	return 0;
+err:
+	kfree(host);
+	return ret;
+}
+
+static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_dai *dai = swidget->private;
+	struct sof_ipc4_dai *ipc4_dai;
+	int node_type = 0;
+	int ret;
+
+	ipc4_dai = kzalloc(sizeof(*ipc4_dai), GFP_KERNEL);
+	if (!ipc4_dai)
+		return -ENOMEM;
+
+	ret = sof_ipc4_get_audio_fmt(scomp, swidget, &ipc4_dai->available_fmt, true);
+	if (ret != 0)
+		goto finish;
+
+	ret = sof_update_ipc_object(scomp, &ipc4_dai->copier.gtw_cfg,
+				    SOF_IPC4_COPIER_GATEWAY_CFG_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(ipc4_dai->copier.gtw_cfg), 1);
+
+	if (ret != 0 || !ipc4_dai->copier.gtw_cfg.dma_buffer_size) {
+		dev_err(scomp->dev, "Invalid dai dma_buffer_size %d\n",
+			ipc4_dai->copier.gtw_cfg.dma_buffer_size);
+		ret = -EINVAL;
+		goto finish;
```

Comment:
finish sets the dai->private. Why do we need to set it in case of error here?

Reply:
> finish sets the dai->private. Why do we need to set it in case of error here?

ack. "finish" label Removed .